File: 198_House_Robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int a = 0, b = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (i % 2 == 0)
                a = max(a + nums[i], b);
            else
                b = max(b + nums[i], a);
        }
        return max(a, b);
    }
};


class Solution {
public:
    int rob(vector<int>& nums) {
        int a = 0, b = 0, temp;
        for (int i = 0; i < nums.size(); i++) {
            temp = a > b ? a : b;
            a = b + nums[i];
            b = temp;
        }
        return  a > b ? a : b;
    }
};

File: 198_House_Robber.py
# Recursion (TLE)
class Solution:
    def rob(self, nums: List[int]) -> int:
        def rob(i):
            if i < 0: return 0
            return max(rob(i - 1), rob(i - 2), nums[i])
        return rob(len(nums) - 1)


# Recursion with memo
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        memo = [-1] * n
        def rob(i):
            if i < 0: return 0
            if memo[i] >= 0: return memo[i]
            res = max(rob(i - 2) + nums[i], rob(i - 1))
            memo[i] = res
            return res
        return rob(n - 1)


# Iteration with memo
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 0: return 0
        memo = [0] * (n + 1)
        memo[1] = nums[0]
        for i in range(1, n):
            memo[i + 1] = max(memo[i], memo[i - 1] + nums[i])
        return memo[-1]


# Iteration with O(1) space
class Solution:
    def rob(self, nums: List[int]) -> int:
        a = b = 0
        for n in nums:
            a, b = max(b + n, a), a
        return a
